### PR TITLE
DPC 1299 implementor org endpoint POST v2/Implementer/:id/org

### PR DESCRIPTION
**Pull request by MrMorie**
### Fixes [DPC-1299](https://jira.cms.gov/browse/DPC-1299)

We need to represent the relationship between implementers and provider organizations. In the web app, implementers will request a relationship with a provider by giving an type 2 NPI. The web app will register this request in the API with a POST /implementer/[id]/org request. In production, the provider org will (eventually) log into a separate portal to approve this request. In sandbox, however, these requests should be automatically approved.

Considerations:

If the NPI does not exist in our organization database, a new organization with that NPI should be created.
In production, we hope to integrate with the NPI database to validate the NPI and find the name of the provider org. In sandbox, a fake name should be used (work with the H-team on naming convention).
There's a many:many relationship between implementers:provider orgs, so we'll need a new table. This table should also describe the status of a relationship (approved/not approved)--perhaps with a column for an approval date.
AC:

In sandbox, a call to POST /implementer/[id]/org will
contain a body with a unique identifier for the org (NPI)
create an org if it does not exist (generate a name at this time)
create an entry in the organizations_implementers  table
in sandbox, set this relationship as approved
respond with success/failure
We have migration scripts to create/remove necessary tables/columns
This feature has appropriate unit tests
### Change Details

- Added /v2/ implementer_org endpoint
- added random name generator for org names
- created migration script to add the new implementor x org relation

### Security Implications

- [ ] new software dependencies

- [ ] security controls or supporting software altered

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

- [x] no PHI/PII is affected by this change

### Feedback Requested

Screen Shot
Screen Shot

### CODE REVIEW

File: dpc-go/dpc-attribution/src/v2/implementer_org.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+// Post function that saves the Implementer to the database and logs any errors before returning a generic error
+func (ios *ImplementerOrgService) Post(w http.ResponseWriter, r *http.Request) {
+	log := logger.WithContext(r.Context())
+
+	implId := r.Context().Value(middleware.ContextKeyImplementer).(string)
+	impl, err := ios.implRepo.FindByID(r.Context(), implId)
+	if err != nil {
+		log.Error("Failed to retrieve Implementer")
```
Comment (MrBilnon) [inline dpc-go/dpc-attribution/src/v2/implementer_org.go:43]: Always good to include the original error in the log too for easier debugging

Comment (MrMorie) [reply dpc-go/dpc-attribution/src/v2/implementer_org.go]: Fixed

File: dpc-go/dpc-attribution/src/model/implementer_org_relation.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"database/sql"
+	"encoding/json"
+	"time"
+)
+
+// ImplementerOrgRelation is a struct that models the Implementer_org_relation table
+type ImplementerOrgRelation struct {
+	ID             string        `db:"id" json:"id"`
+	ImplementerID  string        `db:"implementer_id" json:"implementor_id"`
```
Comment (dhgreene) [inline dpc-go/dpc-attribution/src/model/implementer_org_relation.go:12]:
```suggestion
	ImplementerID  string        `db:"implementer_id" json:"implementer_id"`
```

File: dpc-go/dpc-attribution/src/repository/implementor_org_repository.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
+package repository
```
Comment (dhgreene) [inline dpc-go/dpc-attribution/src/repository/implementor_org_repository.go:1]: Let's rename to file to `implementer_org_repository.go`

File: dpc-go/dpc-attribution/src/v2/implementer_org.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+	adjI, _ := rand.Int(rand.Reader, big.NewInt(int64(len(adj))))
+	nounI, _ := rand.Int(rand.Reader, big.NewInt(int64(len(adj))))
```
Comment (dhgreene) [inline dpc-go/dpc-attribution/src/v2/implementer_org.go:134]:
```suggestion
	nounI, _ := rand.Int(rand.Reader, big.NewInt(int64(len(nouns))))
```

File: dpc-go/dpc-attribution/src/v2/implementer_org.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+		if rel != nil {
+			log.Error("relation already exists", zap.Error(err))
+			boom.Conflict(w, "relation already exists")
+			return
+		}
+	}
+
+	ior, err := ios.impOrgRepo.Insert(r.Context(), implID, org.ID, model.Active)
+	if err != nil {
+		log.Error("Failed to create Implementer org relation", zap.Error(err))
```
Comment (dhgreene) [inline dpc-go/dpc-attribution/src/v2/implementer_org.go:111]: Are we starting errors upper- or lowercase?  Historically, the Go community has wanted lowercase.  I'm seeing a lot of both.

**Review by dhgreene — COMMENTED** [review]
Getting very close!

**Review by dhgreene — APPROVED** [review]
💯

**Pull request merged**